spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -281,35 +297,84 @@ public void onPartitionsAssigned(Map<TopicPartition, Long> assignments, Consumer
 		public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment, Consumer<?, ?> consumer) {
 			Message<?> message = null;
 			try {
-				message = enhanceHeaders(toMessagingMessage(record, acknowledgment, consumer), record);
-				setAttributesIfNecessary(record, message);
+				message = toMessagingMessage(record, acknowledgment, consumer);
 			}
 			catch (RuntimeException e) {
+				if (KafkaInboundGateway.this.retryTemplate == null) {
+					message = enhanceHeaders(message, record);
+				}
 				MessageChannel errorChannel = getErrorChannel();
 				if (errorChannel != null) {
 					KafkaInboundGateway.this.messagingTemplate.send(errorChannel, buildErrorMessage(null,
 							new ConversionException("Failed to convert to message", record, e)));
 				}
 			}
 			if (message != null) {
-				try {
-					Message<?> reply = sendAndReceiveMessage(message);
-					if (reply != null) {
-						reply = enhanceReply(message, reply);
-						KafkaInboundGateway.this.kafkaTemplate.send(reply);
-					}
-				}
-				finally {
-					if (KafkaInboundGateway.this.retryTemplate == null) {
-						ATTRIBUTES_HOLDER.remove();
-					}
-				}
+				sendAndReceive(record, message, acknowledgment, consumer);
 			}
 			else {
 				KafkaInboundGateway.this.logger.debug(() -> "Converter returned a null message for: " + record);
 			}
 		}
 
+		private void sendAndReceive(ConsumerRecord<K, V> record, Message<?> message, Acknowledgment acknowledgment,
+				Consumer<?, ?> consumer) {
+
+			RetryTemplate template = KafkaInboundGateway.this.retryTemplate;
+			if (template != null) {
+				doWithRetry(template, KafkaInboundGateway.this.recoveryCallback, record, acknowledgment, consumer,
+						() -> {
+							Message<?> toSend = enhanceHeaders(message, record, acknowledgment, consumer);
+							if (toSend != null) {
```

Comment:
Don't we need to do anything in the `else`? Like exception or though...
Might be some questions like "where is my message?!"

This might be an existing logic though, therefore I won't mind if we revise it in the separate issue/PR...

Reply:
Turns out the try/catch in enanceHeaders is not necessary - artifact of refactoring; will fix.